dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/InferredEntriesHelper.java

```diff
@@ -0,0 +1,221 @@
+        // which paths point at the primary descriptor of an entry.
+        for (DescriptorLanguage language: getLanguages()) {
+
+            // Determine the list of paths that could be a primary descriptor.
+            Predicate<String> isPrimary = getPrimaryPredicate(language);
+            List<String> paths = allPaths.stream().filter(isPrimary).toList();
+            paths.forEach(p -> LOG.error("POSSIBLE PRIMARY PATH " + p));
+
+            // Remove paths that probably correspond to tests, subtasks, etc.
+            paths = removeTestPaths(paths);
+            paths.forEach(p -> LOG.error("AFTER removeTestPaths PATH " + p));
+
+            // Remove paths that are referenced from other descriptors.
+            paths = removeReferencedPaths(fileTree, paths);
+            paths.forEach(p -> LOG.error("PRIMARY PATH " + p));
+
+            // For each path that remains, calculate its entry type.
+            // If we're successful, add it as an inferred entry.
+            for (String path: paths) {
+                EntryType type = calculateType(fileTree, path, language);
+                String name = calculateName(fileTree, path, language);
+                LOG.error("NAME AND TYPE " + name + " " + type + " " + path + " " + language);
+                if (type != null) {
+                    entries.add(new InferredEntry(type, language, path, name));
+                }
+            }
+        }
+
+        // Postprocess the list of inferred entries to do things like make sure
+        // the entry names are legal.
+        return postprocessNames(entries);
+    }
+
+    private List<DescriptorLanguage> getLanguages() {
+        return List.of(
+            DescriptorLanguage.CWL,
+            DescriptorLanguage.WDL,
+            DescriptorLanguage.NEXTFLOW,
+            DescriptorLanguage.GXFORMAT2,
+            DescriptorLanguage.JUPYTER
+        );
+    }
+
+    private Predicate<String> getPrimaryPredicate(DescriptorLanguage language) {
+        switch (language) {
+        case CWL:
+            return s -> s.endsWith(".cwl");
+        case WDL:
+            return s -> s.endsWith(".wdl");
+        case NEXTFLOW:
+            return s -> s.endsWith("/nextflow.config");
+        case GXFORMAT2:
+            return s -> s.endsWith(".ga") || s.endsWith(".ga2");
+        case JUPYTER:
+            return s -> s.endsWith(".ipynb");
+        default:
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private List<String> removeTestPaths(List<String> paths) {
+        return paths.stream().filter(path -> {
+            String lower = path.toLowerCase();
+            return !(lower.contains("test") || lower.contains("archive") || lower.contains("debug"));
+        }).toList();
+    }
+
+    private List<String> removeReferencedPaths(FileTree fileTree, List<String> paths) {
+        Pattern possiblePath = Pattern.compile("[a-zA-Z0-9_/-]+\\.[a-zA-Z0-9]+");
```

Comment:
I started to try to debug this code to answer this, but then decided to just ask instead... Will this method, including the regex, correctly handle relative path imports? I'm guessing yes